Download all local authority links

AlanGabbianelli opened the pull request:
Allow download of links with specific statuses

The previous behaviour of clicking a link to download broken links has
been modified so that the user can select specific statuses for which
they want to download links. Only links with the selected status will be
present in the downloaded csv file. They are initially all selected by
default.

See specific commit messages for more context on other changes.

Trello: https://trello.com/c/Z6SmqxE3/938-add-button-to-download-all-links-for-a-council

Commit:
Add better_errors gem

Commit:
Add config to avoid having to restart server

In development mode, having to restart the server every time we make a
change is annoying and time consuming. This config will change the
default behaviour.

See: 
https://stackoverflow.com/questions/18270945/rails-server-needs-restart-every-time-i-make-changes-why

Commit:
Add one scope for each link status

And rename currently_broken to broken

Commit:
Fix flaky test by making order irrelevant

thomasleese commented on app/assets/javascripts/local_authorities.js:
```diff
@@ -0,0 +1,15 @@
+        params.push(checkboxes[i].name + '=' + checkboxes[i].value + '&')
+      }
+    }
+    links_download_button.href = href + '?' + params.join('')
```
I think if we do `params.join('&')` here, we can avoid having to do `+ '&'` in the line above which will lead to there being a trailing `&`.

thomasleese commented on app/assets/javascripts/local_authorities.js:
```diff
@@ -0,0 +1,15 @@
+  var links_download_button = document.getElementById('links_download_button');
+  var href = links_download_button.href;
+
+  links_download_button.addEventListener('mouseenter', function() {
```
I wonder if it's better to change the href each time a checkbox gets checked/unchecked, because I'm not sure if this event fires say if a user uses a keyboard to navigate to the button.

AlanGabbianelli replied:
My initial thought was the same and I started implementing it that way, but talking with @edwardkerry I decided to do it this way so to add it only once instead of once for each checkbox.

That said, I see your point, and I think it a stronger one. I'll amend this now.

edwardkerry replied:
Could we add a second eventListener for `onfocus` to catch that, which I think would capture keyboard navigation? It is a small and quick script but might be good to limit how often it runs if possible

AlanGabbianelli replied:
Thinking about it, I feel that, semantically, it's more accurate to add event listeners to the checkboxes. Checking/unchecking a checkbox (not clicking the button) is what changes the outcome.

edwardkerry replied:
Sounds sensible :+1:

thomasleese commented on spec/rails_helper.rb:
```diff
@@ -12,6 +12,8 @@
 require 'capybara/rails'
 require 'capybara/rspec/matchers'
 
+Capybara.server = :puma, { Silent: true }
```
I don't think we need this anymore now that we're using `govuk_test`.

AlanGabbianelli replied:
We do if we want to silence Puma logs and prevent it from writing
```
Puma starting in single mode...
* Version 3.10.0 (ruby 2.4.1-p111), codename: Russell's Teapot
* Min threads: 0, max threads: 1
* Environment: test
* Listening on tcp://0.0.0.0:36775
Use Ctrl-C to stop
```
while running the test suite. See https://github.com/rspec/rspec-rails/issues/1897

AlanGabbianelli replied:
...unless we modify `govuk_test` and put this directly there.

thomasleese replied:
Ah I see! Yeah, I think that sounds like something worth adding to `govuk_test`.

AlanGabbianelli replied:
I'll remove it from here and I'll add it to `govuk_test`

AlanGabbianelli replied:
https://github.com/alphagov/govuk_test/pull/16

AlanGabbianelli replied:
https://github.com/alphagov/govuk_test/pull/17

Commit:
Use updated govuk_test version

This versions silences Puma logs during spec suite:
https://github.com/rspec/rspec-rails/issues/1897

Commit:
Allow download of links with specific statuses

The previous behaviour of clicking a link to download broken links has
been modified so that the user can select specific statuses for which
they want to download links. Only links with the selected status will be
present in the downloaded csv file. They are initially all selected by
default.
